packages/remix-dev/cli/migrate/migrations/replace-remix-imports/getTransformOptions.ts

```diff
@@ -1,30 +1,74 @@
+import inquirer from "inquirer";
 import type { PackageJson } from "type-fest";
 
+import { error, hint } from "../../../../logging";
 import type { Options } from "./transform";
-import { adapters, runtimes } from "./transform";
+import { runtimes, isRuntime, isAdapter } from "./transform";
 import type {
   Adapter,
   Runtime,
 } from "./transform/mapNormalizedImports/packageExports";
 
-const getAdapter = ({ dependencies }: PackageJson): Adapter =>
-  Object.keys(dependencies || {})
-    .filter(
-      (key) =>
-        key.startsWith("@remix-run/") &&
-        !["react", "serve", "server-runtime", ...runtimes]
-          .map((pkgName) => `@remix-run/${pkgName}`)
-          .includes(key)
-    )
-    .map((key) => key.replace("@remix-run/", ""))
-    .find((key) =>
-      adapters.includes(key as typeof adapters[number])
-    ) as Adapter;
-
-const getRuntime = ({ scripts }: PackageJson): Runtime =>
-  (scripts?.postinstall?.replace("remix setup ", "") as Runtime) || "node";
-
-export const getTransformOptions = (packageJson: PackageJson): Options => ({
-  adapter: getAdapter(packageJson),
-  runtime: getRuntime(packageJson),
+const resolveRuntime = async ({ scripts }: PackageJson): Promise<Runtime> => {
+  // match `remix setup <runtime>` in `postinstall` script
+  let remixSetupMatch = scripts?.postinstall?.match(/remix setup (\w+)/);
+  if (remixSetupMatch && remixSetupMatch.length >= 2) {
+    let postinstallRuntime = remixSetupMatch[1];
+    if (isRuntime(postinstallRuntime)) {
+      return postinstallRuntime;
+    }
+  }
+
+  // otherwise, ask user for runtime
```

Comment:
One _tiny_ annoyance is that if users remove `remix setup` from the `postinstall` script, we can no longer auto-detect the runtime reliably, as the runtime is often not a direct dependency of the app.

So for most users, as they re-run `migrate`, we will keep auto-detecting their runtime, until they do what they are supposed to and remove the `remix setup` from `postinstall` script. At that point, the user will have to specify their runtime.

If anyone has a nice solution for this, I'm all ears. For now, I think this is fine, but could _maybe_ be better.

Reply:
I think it's a tradeoff we can't really have otherwise, so don't mind it will be semi-"broken" when people re-run the migration.

Migrations aren't really meant to be run multiple times anyways 🤷‍♂️

Reply:
I think it's a tradeoff we can't really have otherwise, so don't mind it will be semi-"broken" when people re-run the migration.

Migrations aren't really meant to be run multiple times anyways 🤷‍♂️

Reply:
I do like the workflow of running the migration, doing some of what it says, re-running, etc... until everything is green. So in this workflow, `migrate` would ask for the runtime at least once.